### ISSUE TITLE
Map raw json string to correct type?

### ISSUE THREAD

**Issue by davnicwil** [vojtechhabarta/typescript-generator#404]
Hi, firstly thanks very much for this amazing library, it's so useful!

I have come across a situation where I'm storing raw JSON in a database and I'm just holding it in a `String` in my model object to avoid doing the work of mapping the raw JSON to java classes etc. I don't need to do anything with it on the server, just pass it through to the client where it'll be used.

Now, that means typescript-generator by default types these fields as `string`, when in fact they are something like `Array<MyCustomObject>`. Since I "know" they will always be `Array<MyCustomObject>` on the client, is there any way to tell typescript-generator to just give these fields this type in the generated type definitions?

I'm thinking something like an annotation 

```
@WithTypescriptType(type="Array<MyCustomObject>") String rawJsonField;
```

But I can't find anything like this in the docs. Is there a way to do it?

As I say, as a workaround, I can just map the `String` to a `List<MyCustomObject>` in the Java using Jackson or something, but this feels very wasteful if I'm only doing it to generate the client Typescript types correctly. I'd almost rather just modify the generated types manually (or with a script) but this also feels cumbersome/hacky.

Cheers!

**Comment by vojtechhabarta** [comment]
Hi, currently there is no such possibility.

How is this raw JSON transferred? Is it escaped in one JSON property like this:
``` json
{
    "rawJsonField": "[ { \"myCustomObjectField1\": \"\" } ]"
}
```
or is it included naturally like this:
``` json
{
    "rawJsonField": [
        {
            "myCustomObjectField1": ""
        }
    ]
}
```
or do you send just the array like this:
``` json
[
    {
        "myCustomObjectField1": ""
    }
]
```

Do you have sample class where `rawJsonField` is used?

Maybe you could change `rawJsonField` type from `String` to something like `MyCustomObjectsJson` and then use `customTypeMappings` parameter to map `MyCustomObjectsJson` to `Array<MyCustomObject>`.

Is it possible for you to change this type in Java or do you need `String`?

**Comment by davnicwil** [comment]
> Maybe you could change rawJsonField type from String to something like MyCustomObjectsJson and then use customTypeMappings parameter to map MyCustomObjectsJson to Array<MyCustomObject>.

Good idea - I was actually thinking of this approach after reading the docs for that.

I might give this a shot and see how clean it ends up being and report back.

Thing I'm a bit concerned about is that I'm doing the `String` -> literal JSON mapping (i.e. second example in your comment above) using Jackson's `@JsonRawValue` annotation. Without investigating I'm not sure how easily I'll be able to reuse that in my own Jackson type mapper. If I can do that easily, it might be the best way.

If I find a bit of time, I'd also be very happy to take a shot at trying to implement this feature as a PR though - I am thinking something like a `@TypescriptType("List<Foo>")` annotation just overriding whatever the Java type is.

**Comment by vojtechhabarta** [comment]
@davnicwil have you solved it somehow? Could I close this issue?

**Comment by davnicwil** [comment]
Hey @vojtechhabarta sorry was just waiting for a response on

> If I find a bit of time, I'd also be very happy to take a shot at trying to implement this feature as a PR though - I am thinking something like a @TypescriptType("Array<Foo>") annotation just overriding whatever the Java type is.

Like I say, happy to do so & contribute back if you're happy with the shape of that solution

For now I'm still just using my `sed` workaround on the generated types @ build ;-)

**Comment by vojtechhabarta** [comment]
Aha sorry, I was thinking that you wanted to try my suggestion to change String type to something else and use `customTypeMappings` parameter.

I would prefer if you could solve it without adding new feature to typescript-generator but if it is not possible for some reason we could think about adding new parameters like `customTypeAnnotation` and `customTypeAnnotationElement`.

**Comment by davnicwil** [comment]
Ah, right, sorry I didn't get back on that.

So what you suggested works! 

It's a little kludgy with Jackson, to be honest, but it can be done without doing any unnecessary mappings beyond what would be done with `@JsonRawValue`

Here's how I'm doing it, in case others land on this issue and want to save some time:

tl;dr, this maps a class `CustomFieldMapping` -> `Something[]` in the TS types. `CustomFieldMapping` just wraps a `String` of raw JSON.

```Java
// custom serialiser that does the same thing as @JsonRawValue
// but just on a wrapped String field
public class TypescriptGeneratorCustomTypeMapping
    extends JsonSerializer<TypescriptGeneratorCustomTypeMapping.Json> {
  public void serialize(Json json, JsonGenerator jsonGenerator, SerializerProvider provider)
      throws IOException {
    jsonGenerator.writeRawValue(json.rawValue);
  }

  static class Json {
    public String rawValue;
  }
}

// the class we'll use in the typescript-generator mapping
//
// it's just a marker data-holder class, only exists for the mapping
public class CustomFieldMapping extends TypescriptGeneratorCustomTypeMapping.Json {}

// the class we want to generate types for
public class Example {
    
  // this annotation replaces @JsonRawValue
  // and the type is the custom type from above
  @JsonSerialize(using = TypescriptGeneratorCustomTypeMapping.class)
  public CustomFieldMapping something; 

  public Example (String rawJsonValue) {
    this.something = new CustomFieldMapping();
    something.rawValue = rawJsonValue;
  }
}

// then, in the mappings
<mapping>org.foo.bar.CustomFieldMapping:Something[]</mapping>
```

Thanks for this suggestion. Being totally honest, I still think an annotation would be an improvement on this just to be a bit simpler to read and understand what's going on, but the solution above is totally fine, and I get not wanting more features and bug surface are in the lib. Thanks once again for this library, really useful!

**Comment by vojtechhabarta** [comment]
Thanks for very quick response and sharing your solution!
